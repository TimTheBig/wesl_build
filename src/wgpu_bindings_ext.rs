#![cfg(feature = "wgpu_bindings")]

use std::io::{BufWriter, Write};
use std::path::PathBuf;
use std::{fs, path::Path};

use wesl::ModulePath;
use wesl::Mangler;
use wgsl_to_wgpu::WriteOptions;

use crate::WeslBuildExtension;

/// Generate bindings for your wgsl/wesl with wgpu_to_wgsl
///
/// Note this will set the `ManglerKind` to `Escape`
pub struct WgpuBindingsExtension<W: Write> {
    /// The path to output the rust bindings for shaders
    binding_root_path: &'static str,
    /// The courrent module
    bindings_mod_file: W,
    /// The courrent modules path
    bindings_mod_path: PathBuf,
}

impl WgpuBindingsExtension<BufWriter<fs::File>> {
    pub fn new(binding_root_path: &'static str) -> Result<Self, std::io::Error> {
        let bindings_mod_path = Path::new(binding_root_path).join("mod.rs");
        println!("root: {}", bindings_mod_path.display());

        Ok(Self {
            binding_root_path,
            bindings_mod_file: BufWriter::new(fs::File::create(
                &bindings_mod_path,
            )?),
            bindings_mod_path,
        })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum WgpuBindingsError {
    #[error(transparent)]
    IoErr(#[from] std::io::Error),
    #[error(transparent)]
    CreateBindingsModuleErr(#[from] wgsl_to_wgpu::CreateModuleError),
}

impl<WeslResolver: wesl::Resolver> WeslBuildExtension<WeslResolver> for WgpuBindingsExtension<BufWriter<fs::File>> {
    fn name<'n>(&self) -> std::borrow::Cow<'n, str> {
        "WgpuBindingsExtension".into()
    }

    fn init_root(
        &mut self,
        _shader_path: &str,
        res: &mut wesl::Wesl<WeslResolver>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        res.set_mangler(wesl::ManglerKind::Escape);

        writeln!(self.bindings_mod_file, "#![allow(unused)]\n")?;

        Ok(())
    }

    fn into_mod(&mut self, dir_path: &Path) -> Result<(), Box<dyn std::error::Error>> {
        // println!("base: {}", self.bindings_mod_path.display());

        let dir_name = dir_path.file_stem().unwrap().to_str().expect("mod path must be valid UTF-8");
        writeln!(self.bindings_mod_file, "pub(crate) mod {dir_name};")?;

        // remove last mod
        self.bindings_mod_path.pop();
        // move into {dir_name}/"mod.rs
        self.bindings_mod_path.push(dir_name);
        self.bindings_mod_path.push("mod.rs");

        if let Some(dir_to_mod) = self.bindings_mod_path.parent() {
            fs::create_dir_all(dir_to_mod)?;
        }
        // println!("creating: {}", self.bindings_mod_path.display());
        let sub_bindings_mod_file = BufWriter::new(std::fs::File::create(&self.bindings_mod_path)?);
        self.bindings_mod_file = sub_bindings_mod_file;

        Ok(())
    }

    fn exit_mod(&mut self, dir_path: &Path) -> Result<(), Box<dyn std::error::Error>> {
        // println!("{}", self.bindings_mod_path.display());
        self.bindings_mod_path.pop();
        self.bindings_mod_path.pop();
        self.bindings_mod_path.push("mod.rs");
        // println!("{}", self.bindings_mod_path.display());

        self.bindings_mod_file = BufWriter::new(std::fs::File::open(&self.bindings_mod_path)?);

        Ok(())
    }

    fn post_build(&mut self, mod_path: &ModulePath, wgsl_source_path: &str) -> Result<(), Box<dyn std::error::Error>> {
        generate_bindings(
            self.binding_root_path, &mut self.bindings_mod_file,
            mod_path, wgsl_source_path
        ).map_err(|e| Box::<_>::from(e))
    }
}


#[cfg(feature = "wgpu_bindings")]
fn generate_bindings(
    binding_root_path: &str,
    bindings_mod_file: &mut impl Write,
    mod_path: &ModulePath,
    wgsl_source_path: &str,
) -> Result<(), WgpuBindingsError> {
    use wgsl_to_wgpu::MatrixVectorTypes;

    let wgsl_source = std::fs::read_to_string(&wgsl_source_path).unwrap();

    // Configure the output based on the dependencies for the project
    let options = WriteOptions {
        derive_bytemuck_vertex: true,
        derive_encase_host_shareable: true,
        matrix_vector_types: MatrixVectorTypes::Nalgebra,
        ..Default::default()
    };

    // Generate the bindings
    let text = create_shader_module(
        &wgsl_source,
        &wgsl_source_path,
        options
    ).unwrap();

    let binding_path = format!(
        "{}/{}.rs",
        binding_root_path.to_owned(),
        mod_path.components.join("/")
    );
    let binding_path = PathBuf::from(binding_path);

    std::fs::create_dir_all(binding_path.parent().unwrap())?;
    std::fs::write(&binding_path, text.as_bytes())?;

    // Add entry to `mod.rs`
    writeln!(
        bindings_mod_file,
        "pub(crate) mod {};",
        binding_path.file_stem().unwrap().to_str().unwrap()
    )?;

    Ok(())
}

fn create_shader_module(
    wgsl_source: &str,
    wgsl_include_path: &str,
    options: WriteOptions,
) -> Result<String, wgsl_to_wgpu::CreateModuleError> {
    let mut root = wgsl_to_wgpu::Module::default();
    root.add_shader_module(
        wgsl_source,
        Some(wgsl_include_path),
        options,
        wgsl_to_wgpu::ModulePath::default(),
        demangle_wesl,
    )?;
    Ok(root.to_generated_bindings(options))
}

#[cfg(feature = "wgpu_bindings")]
fn demangle_wesl(name: &str) -> wgsl_to_wgpu::TypePath {
    // todo handle `super` paths
    // Assume all paths are absolute paths.
    if name.starts_with("package_") {
        // Use the root module if unmangle fails.
        let mangler = wesl::EscapeMangler;
        let (path, name) = mangler
            .unmangle(name)
            .unwrap_or((ModulePath::new_root(), name.to_string()));

        // Assume all wesl paths are absolute paths.
        wgsl_to_wgpu::TypePath {
            parent: wgsl_to_wgpu::ModulePath {
                components: path.components,
            },
            name,
        }
    } else {
        // Use the root module if the name is not mangled.
        wgsl_to_wgpu::TypePath {
            parent: wgsl_to_wgpu::ModulePath::default(),
            name: name.to_string(),
        }
    }
}
