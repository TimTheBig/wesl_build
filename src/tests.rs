use std::{fs::read_to_string, iter::once};

use insta::assert_snapshot;
use itertools::Itertools;

use super::*;
use crate::wgpu_bindings_ext::WgpuBindingsExtension;

#[test]
fn test_bindings_ext() {
    std::fs::create_dir_all("./test/src/shader_bindings").unwrap();

    #[cfg(feature = "logging")]
    init_build_logger();

    build_shader_dir(
        "./test/src/shaders",
        &mut [Box::new(WgpuBindingsExtension::new("./test/src/shader_bindings").unwrap())]
    ).unwrap();

    // shaders
    assert!(std::fs::exists("test/src/shaders/test.wgsl").unwrap(), "test_bindings_ext requires shaders/test.wgsl to exist");
    assert!(std::fs::exists("test/src/shaders/test2.wgsl").unwrap(), "test_bindings_ext requires shaders/test2.wgsl to exist");

    // bindings
    assert!(std::fs::exists("test/src/shader_bindings/mod.rs").unwrap(), "no mod.rs was generated for shader_bindings root mod");
    assert!(std::fs::exists("test/src/shader_bindings/test_mod/mod.rs").unwrap(), "no mod.rs was generated for shader_bindings test_mod mod");

    assert!(std::fs::exists("test/src/shader_bindings/test.rs").unwrap(), "no shader_binding was generated for test.wgsl");
    assert!(std::fs::exists("test/src/shader_bindings/test2.rs").unwrap(), "no shader_binding was generated for test2.wgsl");
    assert!(std::fs::exists("test/src/shader_bindings/test_mod/test_mod_file.rs").unwrap(), "no shader_binding was generated for test_mod_file.wgsl");

    let mut settings = insta::Settings::new();
    // todo use
    // ("SOURCE", |val, path| {
    //     let source_path = Path::new(val.as_str().unwrap());
    //     assert_eq!(source_path.extension().unwrap(), "wgsl");
    //     assert!(source_path.starts_with(std::env::home_dir().unwrap()));

    //     "source path"
    // });

    let test_bindings = ["test/src/shader_bindings/test.rs", "test/src/shader_bindings/test2.rs", "test/src/shader_bindings/test_mod/test_mod_file.rs"];

    for binding in test_bindings {
        settings.set_input_file(binding);
        settings.bind(|| assert_snapshot!(
            read_to_string(binding).unwrap()
                // skip `SOURCE` const
                .lines().skip(3)
                .interleave_shortest(once("\n").cycle())
                .collect::<String>()
        ));
    }
}
