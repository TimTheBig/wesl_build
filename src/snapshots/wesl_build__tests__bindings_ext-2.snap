---
source: src/tests.rs
expression: "read_to_string(binding).unwrap().lines().skip(3).interleave_shortest(once(\"\\n\").cycle()).collect::<String>()"
input_file: test/src/shader_bindings/test2.rs
---
pub fn create_shader_module(device: &wgpu::Device) -> wgpu::ShaderModule {
    let source = std::borrow::Cow::Borrowed(SOURCE);
    device
        .create_shader_module(wgpu::ShaderModuleDescriptor {
            label: None,
            source: wgpu::ShaderSource::Wgsl(source),
        })
}
pub fn create_pipeline_layout(device: &wgpu::Device) -> wgpu::PipelineLayout {
    device
        .create_pipeline_layout(
            &wgpu::PipelineLayoutDescriptor {
                label: None,
                bind_group_layouts: &[],
                push_constant_ranges: &[],
            },
        )
}
